src/messages/message.ts

```diff
@@ -0,0 +1,49 @@
+import type { JsonMessage } from './types';
+
+import lodash from 'lodash';
+import { validate } from '../validation/validator';
+
+const { cloneDeep, isPlainObject } = lodash;
+export abstract class Message {
+  constructor(protected message: JsonMessage) {}
+
+  static getJsonSchema(): object {
+    throw new Error('method not implemented');
+  }
+  static getType(): string {
+    throw new Error('method not implemented');
+  };
+
+  static unmarshal(rawMessage: object): JsonMessage {
+    const descriptor = rawMessage['descriptor'];
+    if (!descriptor) {
+      throw new Error('message must contain descriptor');
+    }
+
+    if (!isPlainObject(descriptor)) {
+      throw new Error('descriptor: must be object');
+    }
+
+    const messageType = descriptor['method'];
+    if (!messageType) {
+      throw new Error('descriptor must contain method');
+    }
+
+    // validate throws an error if message is invalid
+    validate(messageType, rawMessage);
+
+    return rawMessage as JsonMessage;
+  };
+
+  getMethod(): string {
+    return this.message.descriptor.method;
+  }
+
+  toObject(): JsonMessage {
+    return cloneDeep(this.message);
+  }
+
+  toJSON(): string {
+    return JSON.stringify(this.message);
+  }
```

Comment:
The need for a class here is also so thin: `getMethod()` not used (DWN accesses `JsonMessage`'s `descriptor.method` directly, `toObject()` and `toJson()` can easily pass with static methods or function exports.

I am inclined to favor changing this into just type/interface.

Reply:
how do we provide default implementations of static methods using an type/interface?

Reply:
Can't define interface for static methods. But here I see them as just static method with no requirements to override. Essentially:

```typescript
static getMethod(message: MessageSchema): string {
...
}

static toObject(message: MessageSchema): MessageJson {
...
}


static toJSON(message: MessageSchema): string {
...
}
```

Reply:
I think i'm going to leave these as is for now. feels a bit easier to just call `toObject` on an "instance" of `Message` and `toJSON` overrides default `JSON.stringify` behavior so i don't think we can make it static

Reply:
Don't understand the comment on "default `JSON.stringify` behavior" so not 100% convinced static methods can't be used. At the same time there are better things to argue over for sure, so not losing sleep on this one.

Will leave this open one last time and let it simmer a bit more since there is an earlier comment on `create()` methods related to this class, will resolve both if no additional action comes out of that.

Reply:
* i like being able to call `toObject` like [this](https://github.com/TBD54566975/dwn-sdk-js/blob/refactor/src/store/message-store-level.ts#L129)
* [overriding default `JSON.stringify` behavior](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#tojson_behavior)